### PR TITLE
Add a consistency check function

### PR DESCRIPTION
**Pull request by leplatrem**
https://bugzilla.mozilla.org/show_bug.cgi?id=1521030

- [x] basic idea
- [x] seems to work fine
- [x] skip preview entries
- [ ] document

### CODE REVIEW

**Comment by leplatrem** [comment]
```
$ AUTH=aQwtxo...5zevj SERVER=https://settings-writer.stage.mozaws.net/v1 python aws_lambda.py consistency_checks
blocklists/addons OK
blocklists/certificates OK
blocklists/gfx OK
blocklists/plugins OK
blocklists/qa OK
main/cfr OK
main/chinarepack-newtab-configuration OK
main/chinarepack-newtab-topsites OK
main/fenix-experiments OK
main/fftv-experiments OK
main/focus-experiments OK
main/fxmonitor-breaches OK
main/language-dictionaries OK
main/lite-experiments OK
main/normandy-recipes OK
main/onboarding OK
main/personality-provider-models OK
main/personality-provider-recipe OK
main/product-integrity OK
main/rocket-prefs OK
main/rocket-releases OK
main/sites-classification OK
main/tippytop OK
main/url-classifier-skip-urls OK
pinning/pins OK
security-state/cert-revocations OK
security-state/intermediates OK
security-state/onecrl OK
```

File: aws_lambda.py [hunk 1]
```diff
@@ -318,6 +324,81 @@ def get_signed_source(server_info, change):
...
+            diff = compare_records(source_records, preview_records)
+            if diff:
+                raise ValueError(f"Inconsistency detected: {diff}")
+
+        elif source_metadata["status"] == "signed":
+            preview_records = client.get_records(**r["preview"])
+            dest_records = client.get_records(**r["destination"])
+            diff = compare_records(preview_records, dest_records)
+            if diff:
+                raise ValueError(f"Inconsistency detected: {diff}")
```
Comment (glasserc) [inline aws_lambda.py]: You might want to indicate the bucket/collection where this inconsistency happened.

File: aws_lambda.py [hunk 2]
```diff
@@ -318,6 +324,81 @@ def get_signed_source(server_info, change):
...
+    except KeyError:
+        raise ValueError("No signer capabilities found. Run on *writer* server!")
+
+    signed_collections = {}
+    preview_buckets = set()
+    for resource in resources:
+        if resource["source"]["collection"] is not None:
+            signed_collections[(resource["destination"]["bucket"], resource["destination"]["collection"])] = resource
+        else:
+            signed_collections[(resource["destination"]["bucket"],)] = resource
```
Comment (glasserc) [inline aws_lambda.py]: Why not use two dicts? This one could be called `signed_buckets`.

File: aws_lambda.py [hunk 3]
```diff
@@ -318,6 +324,81 @@ def get_signed_source(server_info, change):
...
+
+        # Skip preview collections entries
+        if bid in preview_buckets:
+            continue
+
+        if (bid, cid) in signed_collections:
+            r = signed_collections[(bid, cid)]
+        elif (bid,) in signed_collections:
+            r = signed_collections[(bid,)]
+            r["source"]["collection"] = r["preview"]["collection"] = r["destination"]["collection"] = cid
```
[the remark this answers is not in the document]
Comment (glasserc) [reply aws_lambda.py]: I would say yes.

File: aws_lambda.py [hunk 4]
```diff
@@ -318,6 +324,81 @@ def get_signed_source(server_info, change):
...
 
 
+def compare_records(a, b):
+    """Compare two lists of records. Returns empty list if equal."""
+    b_by_id = {r["id"]: r for r in b}
+    diff = []
+    for ra in a:
+        rb = b_by_id.pop(ra["id"], None)
+        if rb is None:
+            diff.append(ra)
```
Comment (glasserc) [inline aws_lambda.py]: Check that `ra == rb`?

File: aws_lambda.py [hunk 5]
```diff
@@ -318,6 +324,81 @@ def get_signed_source(server_info, change):
...
+            if diff:
+                raise ValueError(f"Inconsistency detected: {diff}")
+
+        elif source_metadata["status"] == "signed":
+            preview_records = client.get_records(**r["preview"])
+            dest_records = client.get_records(**r["destination"])
+            diff = compare_records(preview_records, dest_records)
+            if diff:
+                raise ValueError(f"Inconsistency detected: {diff}")
+
```
Comment (glasserc) [inline aws_lambda.py]: Maybe

```
else:
    print("{bucket}/{collection} status {status}, no assertions are possible".format(**source_metadata))
```

?

File: aws_lambda.py (hunk 1 above)
Comment (glasserc) [reply aws_lambda.py]: Also, I think you could check `r["source"]` against either `r["preview"]` or `r["destination"]`.

File: aws_lambda.py [hunk 6]
```diff
@@ -318,6 +325,134 @@ def get_signed_source(server_info, change):
...
+    # For most cases, configuration of signed resources is specified by bucket and
+    # does not contain any collection information.
+    resources_by_bid = {}
+    resources_by_cid = {}
+    preview_buckets = set()
+    for resource in resources:
+        if resource["source"]["collection"] is not None:
+            resources_by_cid[(resource["destination"]["bucket"], resource["destination"]["collection"])] = resource
+        else:
+            resources_by_bid[(resource["destination"]["bucket"],)] = resource
```
Comment (glasserc) [inline aws_lambda.py]: It isn't really necessary to index them by tuple any more.

**Pull request merged**